### PR TITLE
8200610: Compiling fails with java.nio.file.ReadOnlyFileSystemException

### PR DESCRIPTION
**Pull request by archiecobbs**
In `JavaCompiler`, when trying to write a class file, if an `IOException`s is thrown, instead of crashing, we report a normal error" `error while writing [file]: ...`.

However there are some I/O errors that can generate `RuntimeException`s not caught by this logic. So instead these errors cause a compiler crash.

An example is given in JDK-8200610, where the compiler tries to write into a ZIP file system and a `java.nio.file.ReadOnlyFileSystemException` is thrown.

This patch adds that exception and a couple of others which (in my estimation) could reasonably be expected to be thrown in various obscure cases involving alternative file systems: `FileSystemNotFoundException` and `InvalidPathException`.

Any suggestions on what the most appropriate list of exception types to include here are appreciated.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8200610](https://bugs.openjdk.org/browse/JDK-8200610): Compiling fails with java.nio.file.ReadOnlyFileSystemException

### Reviewers
 * [Vicente Romero](https://openjdk.org/census#vromero) (@vicente-romero-oracle - **Reviewer**) ⚠️ Review applies to [1bcf9501](https://git.openjdk.org/jdk/pull/10818/files/1bcf9501441bcda9c60fe1c6e3b5162c52c632c3)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10818/head:pull/10818` \
`$ git checkout pull/10818`

Update a local copy of the PR: \
`$ git checkout pull/10818` \
`$ git pull https://git.openjdk.org/jdk pull/10818/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10818`

View PR using the GUI difftool: \
`$ git pr show -t 10818`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10818.diff">https://git.openjdk.org/jdk/pull/10818.diff</a>

</details>

### CODE REVIEW

**Comment by jonathan-gibbons** [comment]
If we're hypothesizing plausible unchecked exceptions, maybe consider adding `UncheckedIOException` ?

File: src/jdk.compiler/share/classes/com/sun/tools/javac/main/JavaCompiler.java [hunk 1]
```diff
@@ -1637,7 +1640,7 @@ public void generate(Queue<Pair<Env<AttrContext>, JCClassDecl>> queue, Queue<Jav
                 }
                 if (results != null && file != null)
                     results.add(file);
-            } catch (IOException ex) {
+            } catch (IOException | FileSystemNotFoundException | InvalidPathException | ReadOnlyFileSystemException ex) {
```
Comment (jonathan-gibbons) [inline src/jdk.compiler/share/classes/com/sun/tools/javac/main/JavaCompiler.java:1643]: If we're hypothesizing plausible unchecked exceptions, maybe consider adding UncheckedIOException ?

I'd also recommend one exception per line, breaking the line before the `|`

```
} catch (IOException 
    | FileSystemNotFoundException 
    | InvalidPathException 
    | ReadOnlyFileSystemException 
    | UncheckedIOException ex) {
```

**Review by vicente-romero-oracle — APPROVED** [review]
lgtm

**Comment by archiecobbs** [comment]
Thanks!

/integrate

**Comment by vicente-romero-oracle** [comment]
hi @archiecobbs, it seems like you need to issue the `/integrate` command again for this one. Thanks

**Comment by archiecobbs** [comment]
Bleh, Mr. Bot is going in circles. Now it says you have to `/sponsor` again....

**Comment by vicente-romero-oracle** [comment]
> Bleh, Mr. Bot is going in circles. Now it says you have to `/sponsor` again....

yep :( saw that too, let's wait

**Pull request closed without merging**